File: java/com/LangStack/Task/TaskThreadPool.java
package com.LangStack.Task;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class TaskThreadPool
{
    private static final int         MIN_THREAD_COUNT = 4;      ///< 线程池最少线程
    private static final int         MAX_THREAD_COUNT = 64;     ///< 线程池最大线程
    private static final int         THREAD_IDLE_TIME = 15000;  ///< 线程最大空闲时间

    private static TaskThreadPool    sTaskThreadPool  = null;   ///< 单例
    
    private Map<Integer, TaskThread> mThreadPool      = null;   ///< 线程池
    private Map<String, Integer>     mIdentifyMap     = null;   ///< 任务ID索引
    private Lock                     mLock            = null;   ///< 锁
    private int                      mIndex           = 0;      ///< 内部线程索引
    

    /**
     * @return 单例
     */
    public synchronized static TaskThreadPool instance()
    {
        if (sTaskThreadPool == null)
        {
            sTaskThreadPool = new TaskThreadPool();
        }
        return sTaskThreadPool;
    }
    
    /**
     * @brief       添加任务
     * @param       task        任务
     */
    public void addTask(ITask task)
    {
        mLock.lock();
        getTheLightThread(task.getTaskId()).addTask(task);
        mLock.unlock();
    }
    
    /**
     * @brief       线程空闲回调
     * @param       index       线程索引
     */
    public void onThreadIdle(int index)
    {
        mLock.lock();
        if (mThreadPool.containsKey(index))
        {
            TaskThread thread = mThreadPool.get(index);
            if (thread.getWeight() == 0)
            {
                ITask task = new DeleteThreadTask(thread);
                mThreadPool.remove(index);
                getTheLightThread(task.getTaskId()).addTask(task);
            }
        }
        mLock.unlock();    
    }
    
    /**
     * @biref       线程任务队列空闲回调
     * @param       taskId      任务队列标识
     * @param       index       线程ID
     */
    public void onTaskIdle(String taskId, int index)
    {
        mLock.lock();
        if (mIdentifyMap.containsKey(taskId))
        {
            if (mIdentifyMap.get(taskId) == index)
            {
                mIdentifyMap.remove(taskId);
            }
        }
        mLock.unlock();     
    }
    
    public int getThreadCount()
    {
        mLock.lock();
        int res = mThreadPool.size();
        mLock.unlock();  
        return res;
    }
    
    /**
     * @brief       获取负载最轻的线程
     * @param       taskId          任务ID(相同任务ID会分配同一个线程)
     * @return      线程
     */
    private TaskThread getTheLightThread(String taskId)
    {
        if (mIdentifyMap.containsKey(taskId))
        {
            int key = mIdentifyMap.get(taskId);
            if (mThreadPool.containsKey(key))
            {
                return mThreadPool.get(mIdentifyMap.get(taskId));
            }
        }
        
        int weight = Integer.MAX_VALUE;
        TaskThread lightThread = null;
        for (TaskThread thread : mThreadPool.values())
        {
            if (thread.getWeight() < weight)
            {
                weight = thread.getWeight();
                lightThread = thread;
                if (weight == 0)
                {
                	break;
                }
                
            }
        }
        
        if (weight != 0 && mThreadPool.size() < MAX_THREAD_COUNT)
        {
            lightThread = new TaskThread(this, THREAD_IDLE_TIME, mIndex);
            mThreadPool.put(mIndex, lightThread);
            while (mThreadPool.containsKey(mIndex)) 
            {
                ++mIndex;
            }
        }
        
        mIdentifyMap.put(taskId, lightThread.getIndex());
        return lightThread;
    }

    /**
     * @brief       构造，创建默认的最少线程
     */
    private TaskThreadPool()
    {
        mThreadPool = new HashMap<Integer, TaskThread>();
        mIdentifyMap = new HashMap<String, Integer>();
        mLock = new ReentrantLock();
        for (int i = 0; i < MIN_THREAD_COUNT; ++i)
        {
        	/// 最低线程永不超时
            mThreadPool.put(mIndex, new TaskThread(this, -1, mIndex));
            while (mThreadPool.containsKey(mIndex)) 
            {
                ++mIndex;
            }
        }
    }
}

File: cpp/src/Task/TaskThreadPool.cpp
/******************************************************************************
Copyright (C) 2015 zhuyanxie

　　Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is furnished
to do so, subject to the following conditions:

　　The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
******************************************************************************/


#include "Task/TaskThreadPool.h"

#include "DeleteThreadTask.h"
#include "Log/Log.h"
#include "LangStackConstant.h"

namespace ls {

std::mutex CTaskThreadPool::s_initLock;
CTaskThreadPool* CTaskThreadPool::s_instance = NULL;

CTaskThreadPool* CTaskThreadPool::instance()
{
    if (NULL == s_instance)
    {
        std::lock_guard<std::mutex> lck(s_initLock);
        if (NULL == s_instance)
        {
            s_instance = new CTaskThreadPool();
        }
    }

    return s_instance;
}

CTaskThreadPool::CTaskThreadPool(int minThreadCount, int maxThreadCount,
        int maxIdleTime)
    : m_index(0)
    , m_minCount(minThreadCount)
    , m_maxCount(maxThreadCount)
    , m_idleTime(maxIdleTime)
{
    for (int i = 0; i < minThreadCount; ++i)
    {
        addThread(-1);
    }
}

CTaskThreadPool::~CTaskThreadPool()
{
    m_taskThreads.clear();
}

///\brief			设置新创建的线程的空闲超时时间
void CTaskThreadPool::setThreadIdleTime(int idleTime)
{
	m_idleTime = idleTime <= 0 ? 5000 : idleTime;
}

///\brief           添加一个任务到线程池
void CTaskThreadPool::addTask(ITask* task)
{
    std::lock_guard<std::mutex> lck(m_threadLock);
    getTheLightThread(task->getTaskId())->addTask(task);
}

///\brief           获取负载较轻的线程
TaskThreadPtr CTaskThreadPool::getTheLightThread(const std::string &id)
{
    std::map<std::string, int>::iterator itmap = mIdentifyMap.find(id);
    if (mIdentifyMap.end() != itmap && m_taskThreads.count(itmap->second))
    {
    	VERBOSEF(LS_TAG, "get thread sucess repeat id [%s->%d]\n",
    			id.c_str(), itmap->second);
        return m_taskThreads[itmap->second];
    }

    auto res = m_taskThreads.begin();
    int weight = res->second->getWeight();
    for (auto it = m_taskThreads.begin(); it != m_taskThreads.end(); ++it)
    {
        if (weight == 0)
        {
            break;
        }
        if (weight > it->second->getWeight())
        {
            res = it;
            weight = it->second->getWeight();
        }
    }

    if (weight != 0 && m_taskThreads.size() < size_t(m_maxCount))
    {
        int index = addThread(m_idleTime);
        res = m_taskThreads.find(index);
    }
	DEBUGF(LS_TAG, "get thread sucess [%s->%d]\n", id.c_str(), res->first);
    mIdentifyMap[id] = res->first;
    return res->second;
}

///\brief           在线程池中增加一个线程
int CTaskThreadPool::addThread(int maxIdleTime)
{
    m_taskThreads[m_index] = TaskThreadPtr(
        new TaskThread(this, maxIdleTime, m_index));
   return m_index++;
}

///\brief           线程空闲回调
void CTaskThreadPool::onThreadIdle(int index)
{
    std::lock_guard<std::mutex> lck(m_threadLock);
    auto it = m_taskThreads.find(index);
    if (it != m_taskThreads.end())
    {
        ITask *task = new CDeleteThreadTask(it->second);
        m_taskThreads.erase(index);
        getTheLightThread(task->getTaskId())->addTask(task);
    }
}

///\brief           任务队列完成回调
void CTaskThreadPool::onTaskIdle(const std::string &taskId, int index)
{
    std::lock_guard<std::mutex> lck(m_threadLock);
    auto it = mIdentifyMap.find(taskId);
    if (it->second == index)
    {
        mIdentifyMap.erase(it);
    }
}

}
